lib/collection/src/shards/resharding/driver.rs

```diff
@@ -94,21 +95,41 @@ impl DriverState {
             .for_each(|peer_stage| *peer_stage = next_stage.max(*peer_stage));
     }
 
-    /// List the shard IDs we still need to migrate.
+    /// List the shard ID pairs we still need to migrate
+    ///
+    /// When scaling up this produces shard IDs to migrate points from. When scaling down this
+    /// produces shard IDs to migrate points into.
     pub fn shards_to_migrate(&self) -> impl Iterator<Item = ShardId> + '_ {
-        self.source_shards()
-            .filter(|shard_id| !self.migrated_shards.contains(shard_id))
+        self.shards()
+            // Exclude current resharding shard, and already migrated shards
+            .filter(|shard_id| {
+                *shard_id != self.key.shard_id && !self.migrated_shards.contains(shard_id)
+            })
     }
```

Comment:
We don't need to know about peer IDs here. That would only make the structure more complex with no added value.

In this implementation we send to all peers concurrently so there's no 'stages' here.

That might change however depending on how we revise the transfer logic of this implementation. For example, in <https://github.com/qdrant/qdrant/pull/4741> I did include the peer ID because we send to each replica separately.